src/metabase/models/data_permissions.clj

```diff
@@ -0,0 +1,311 @@
+;;    user is in. This can be overridden by definding a custom implementation of `coalesce`.
+;;
+;;  - If a user does not have any value for the permission when it is fetched, the *least* permissive value is used as a
+;;    fallback.
+
+
+(def ^:private Permissions
+  "Permissions which apply to individual databases or tables"
+  {:data-access           {:model :model/Table :values [:unrestricted :no-self-service :block]}
+   :download-results      {:model :model/Table :values [:one-million-rows :ten-thousand-rows :no]}
+   :manage-table-metadata {:model :model/Table :values [:yes :no]}
+
+   :native-query-editing {:model :model/Database :values [:yes :no]}
+   :manage-database      {:model :model/Database :values [:yes :no]}})
+
+(def PermissionType
+  "Malli spec for valid permission types."
+  (into [:enum {:error/message "Invalid permission type"}]
+        (keys Permissions)))
+
+(def PermissionValue
+  "Malli spec for a keyword that matches any value in [[Permissions]]."
+  (into [:enum {:error/message "Invalid permission value"}]
+        (distinct (mapcat :values (vals Permissions)))))
+
+
+;;; ------------------------------------------- Misc Utils ------------------------------------------------------------
+
+(defn- least-permissive-value
+  "The *least* permissive value for a given perm type. This value is used as a fallback when a user does not have a
+  value for the permission in the database."
+  [perm-type]
+  (-> Permissions perm-type :values last))
+
+(defn- most-permissive-value
+  "The *most* permissive value for a given perm type. This is the default value for superusers."
+  [perm-type]
+  (-> Permissions perm-type :values first))
+
+(def ^:private model-by-perm-type
+  "A map from permission types directly to model identifiers (or `nil`)."
+  (update-vals Permissions :model))
+
+(defn- assert-value-matches-perm-type
+  [perm-type perm-value]
+  (when-not (contains? (set (get-in Permissions [perm-type :values])) perm-value)
+    (throw (ex-info (tru "Permission type {0} cannot be set to {1}" perm-type perm-value)
+                    {perm-type (Permissions perm-type)}))))
+
+
+;;; ---------------------------------------- Fetching a user's permissions --------------------------------------------
+
+(defmulti coalesce
+  "Coalesce a set of permission values into a single value. This is used to determine the permission to enforce for a
+  user in multiple groups with conflicting permissions. By default, this returns the *most* permissive value that the
+  user has in any group.
+
+  For instance,
+  - Given an empty set, we return the most permissive.
+    (coalesce :settings-access #{}) => :yes
+  - Given a set with values, we select the most permissive option in the set.
+    (coalesce :settings-access #{:view :no-access}) => :view"
+  {:arglists '([perm-type perm-values])}
+  (fn [perm-type _perm-values] perm-type))
+
+(defmethod coalesce :default
```

Comment:
I almost didn't mention it, but coalesce is kind of an overloaded word, maybe we can use a totally metabase-unique word for it? Working on `dox` was so easy since it was very easy to jump to the couple of different `dox` names in the code. Don't let it block us but I just wanted to share my thoughts

Reply:
Sure, this is purely internal to this namespace so we can brainstorm a better name. Merging as-is for now.